src/window.c

```diff
@@ -6314,6 +6334,104 @@ set_fraction(win_T *wp)
 				     + FRACTION_MULT / 2) / (long)wp->w_height;
 }
 
+/*
+ * Handle scroll position for 'nosplitscroll', replaces
+ * scroll_to_fraction() call from win_new_height().
+ * Instead we iterate over all windows in a tabpage and
+ * calculate the new scroll/cursor position.
+ * TODO: Ensure this also works with wrapped lines.
+ * Requires topline to be able to be set to a bufferline
+ * with some offset(row-wise scrolling/smoothscroll).
+ */
+    static void
+win_fix_scroll(win_T *next_curwin, int close)
+{
+    int      ls_offset = (p_ls == 1 && win_count() == 2) ? 1 : 0;
+    win_T    *wp;
+    linenr_T lnum;
+
+    FOR_ALL_WINDOWS(wp)
+    {
+	// Skip when window height has not changed or when
+	// buffer has less lines than the window height.
+	if (wp->w_height != wp->w_prev_height
+		&& wp->w_height < wp->w_buffer->b_ml.ml_line_count)
+	{
+	    lnum = wp->w_cursor.lnum;
+	    // Determine cursor position required to keep the same screenlines.
+	    if (wp->w_winrow != wp->w_prev_winrow)
+	    { // For bottom windows, keep the same botline except for 'ls' == 1.
+		if (!close && ((wp->w_winrow + VISIBLE_HEIGHT(wp)
+					+ wp->w_status_height) == cmdline_row))
+		    wp->w_cursor.lnum = wp->w_botline - 1 - ls_offset;
+		else // Otherwise determine based on winrow and height.
+		    wp->w_cursor.lnum = MIN(wp->w_botline - 1
```

Comment:
I think this can evaluate to negative values, do we need to guard that here? I think it is caught in some other function. Don't see any problems at least.